src/openms/source/ANALYSIS/OPENSWATH/SpectrumExtractor.cpp

```diff
@@ -419,23 +438,65 @@ namespace OpenMS
 
     // score spectra
     std::vector<MSSpectrum> scored;
-    scoreSpectrum(annotated, picked, scored);
+    scoreSpectrum(annotated, picked, scored, features);
 
-    // select the best spectrum for each transition of the target list
-    transition_best_spec.clear();
-    for (auto spectrum : scored)
+    selectSpectra(scored, extracted_spectra, features, extracted_features);
+  }
+
+  void SpectrumExtractor::selectSpectra(
+    const std::vector<MSSpectrum>& scored_spectra,
+    std::vector<MSSpectrum>& selected_spectra,
+    const FeatureMap& features,
+    FeatureMap& selected_features
+  )
+  {
+    std::unordered_map<std::string,UInt> transition_best_spec;
+    for (UInt i=0; i<scored_spectra.size(); ++i)
     {
-      String transition_name = spectrum.getName();
-      std::map<std::string,MSSpectrum>::const_iterator it = transition_best_spec.find(transition_name);
+      String transition_name = scored_spectra[i].getName();
+      std::unordered_map<std::string,UInt>::const_iterator it = transition_best_spec.find(transition_name);
       if (it == transition_best_spec.end())
       {
-        transition_best_spec.insert({transition_name, spectrum});
+        transition_best_spec.insert({transition_name, i});
       }
-      else if (it->second.getFloatDataArrays()[1][0] < spectrum.getFloatDataArrays()[1][0])
+      else if (scored_spectra[it->second].getFloatDataArrays()[1][0] < scored_spectra[i].getFloatDataArrays()[1][0])
       {
         transition_best_spec.erase(transition_name);
-        transition_best_spec.insert({transition_name, spectrum});
+        transition_best_spec.insert({transition_name, i});
       }
     }
+
+    for (auto it = transition_best_spec.cbegin(); it!=transition_best_spec.cend(); ++it)
+    {
+      selected_spectra.push_back(scored_spectra[it->second]);
+      selected_features.push_back(features[it->second]);
+    }
+  }
+
+  void SpectrumExtractor::selectSpectra(
+    const std::vector<MSSpectrum>& scored_spectra,
+    std::vector<MSSpectrum>& selected_spectra
+  )
+  {
+    std::unordered_map<std::string,UInt> transition_best_spec;
```

Comment:
This can be simplified by the following:

`void SpectrumExtractor::selectSpectra(
  const std::vector<MSSpectrum>& scored_spectra,
 std::vector<MSSpectrum>& selected_spectra)
{
FeatureMap features;
FeatureMap extracted_features;
selectSpectra(scored_spectra, selected_spectra, features, extracted_features);
}`